src/qr_login.rs

```diff
@@ -0,0 +1,140 @@
+    fn from(value: qr_login::QrCodeMode) -> Self {
+        match value {
+            qr_login::QrCodeMode::Login => Self::Login,
+            qr_login::QrCodeMode::Reciprocate => Self::Reciprocate,
+        }
+    }
+}
+
+/// Data for the QR code login mechanism.
+///
+/// The [`QrCodeData`] can be serialized and encoded as a QR code or it can be
+/// decoded from a QR code.
+#[wasm_bindgen]
+#[derive(Debug)]
+pub struct QrCodeData {
+    inner: qr_login::QrCodeData,
+}
+
+#[wasm_bindgen]
+impl QrCodeData {
+    /// Create new [`QrCodeData`] from a given public key, a rendezvous URL and,
+    /// optionally, a homeserver ULR.
+    ///
+    /// If a homeserver URL is given, then the [`QrCodeData`] mode will be
+    /// [`QrCodeMode::Reciprocate`], i.e. the QR code will contain data for the
+    /// existing device to display the QR code.
+    ///
+    /// If no homeserver is given, the [`QrCodeData`] mode will be
+    /// [`QrCodeMode::Login`], i.e. the QR code will contain data for the
+    /// new device to display the QR code.
+    #[wasm_bindgen(constructor)]
+    pub fn new(
+        public_key: Curve25519PublicKey,
+        rendezvous_url: &str,
+        homeserver_url: Option<String>,
+    ) -> Result<QrCodeData, JsError> {
+        let public_key = public_key.inner;
+        let rendezvous_url = Url::parse(rendezvous_url)?;
+
+        let mode_data = if let Some(homeserver_url) = homeserver_url {
+            qr_login::QrCodeModeData::Reciprocate { homeserver_url: Url::parse(&homeserver_url)? }
+        } else {
+            qr_login::QrCodeModeData::Login
+        };
+
+        let inner = qr_login::QrCodeData { public_key, rendezvous_url, mode_data };
+
+        Ok(QrCodeData { inner })
+    }
+
+    /// Attempt to decode a slice of bytes into a [`QrCodeData`] object.
+    ///
+    /// The slice of bytes would generally be returned by a QR code decoder.
+    pub fn from_bytes(bytes: &[u8]) -> Result<QrCodeData, JsError> {
+        Ok(Self { inner: qr_login::QrCodeData::from_bytes(bytes)? })
+    }
+
+    /// Encode the [`QrCodeData`] into a list of bytes.
+    ///
+    /// The list of bytes can be used by a QR code generator to create an image
+    /// containing a QR code.
+    pub fn to_bytes(&self) -> Vec<u8> {
+        self.inner.to_bytes()
+    }
+
+    /// Attempt to decode a base64 encoded string into a [`QrCodeData`] object.
+    pub fn from_base64(data: &str) -> Result<QrCodeData, JsError> {
+        Ok(Self { inner: qr_login::QrCodeData::from_base64(data)? })
+    }
+
+    /// Encode the [`QrCodeData`] into a list of bytes.
+    ///
+    /// The list of bytes can be used by a QR code generator to create an image
+    /// containing a QR code.
+    pub fn to_base64(&self) -> String {
+        self.inner.to_base64()
+    }
+
+    /// Get the Curve25519 public key embedded in the [`QrCodeData`].
+    ///
+    /// This Curve25519 public key should be used to establish an ECIES channel
```

Comment:
Can you explicit the ECIES acronym once, or point to a doc / link that does?

Reply:
We do link to the wiki page for ECIES in the vodozemac/ecies module, and we deffer to the vodozemac docs to not copy paste everything from vodozemac. With that in mind do you want me to link to the wiki page again here?

Reply:
Ah thanks, maybe we could explicit the acronym once at least?

Reply:
Added the link and defined the acronym: 4be534b3035a8bb96192fdfe6b9c2e656f0d3bef.